71-tgyuuAn

tgyuuAn opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[횡단보도](https://www.acmicpc.net/problem/24042)

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
**1시간 20분**

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->

![image](https://github.com/user-attachments/assets/2c5f5d70-26d9-4fec-9fe0-12cdd3c2555c)

완탐으로 풀었을 때 시간 복잡도 : $100,000 * 700,000 = 7백억$ 당연히 시간 초과!!!!!!

<br><br><br>

`노드 <= 10만`, `간선 <= 70만`인 문제를 풀기위해서는 최대 $N log (N)$의 시간복잡도 이내로 풀어야 함. _(1억번 연산을 넘으면 안되기 때문)_

이 때 문제에서 힌트로 주어진 것이 바로 메모리 제한이 1024MB로 매우 널널하다는 것

<img width="306" alt="image" src="https://github.com/user-attachments/assets/c041fef5-8ac8-443c-8df7-32f82af78ed7">

<br><br><br>

그럼 DP로 접근해야되나 ...? 라고 곰곰히 생각해봤는데 도저히 DP로 풀 수 없다.

그래서 $N*log(N)$ 이내의 알고리즘이 뭐가 있을까 쭉 나열해봤다.

<br><br><Br>

힙, 정렬 + 이분탐색, 정렬, 다익스트라,  ,,,,,,,,,, **다익스트라** ?!

<br><br><br>

다익스트라의 시간복잡도는 $간선의 수 * log(간선의 수)$

= 70만 * log(70만) 

= 70만 * ( log(1024) + log(700) )

= 70만 * (10 + 약 9)

= 1310만

-> 완죤 가능...!

<br><br><br><br><br><br>

그래서 코드를 바로 와다다 썼고,

```python
from collections import defaultdict
from heapq import *
import sys

def input(): return sys.stdin.readline().rstrip()

N, M = map(int, input().split())

graph = defaultdict(lambda : defaultdict(int))

for idx in range(M):
    start, destination = map(int, input().split())
    graph[start][destination] = min(graph[start][destination], idx) if graph[start][destination] != 0 else idx
    graph[destination][start] = min(graph[destination][start], idx) if graph[destination][start] != 0 else idx

costs = [int(1e9) for _ in range(N+1)]
costs[1] = 0
visited = set()
heap = [(0, 1)]
while heap:
    now_cost, now_idx = heappop(heap)
    
    if now_idx in visited: continue
    visited.add(now_idx)
    
    costs[now_idx] = now_cost
    if now_idx == N: break

    for neighbor in graph[now_idx]:
        if graph[now_idx][neighbor] == 0: continue
        if neighbor in visited: continue

        if now_cost + graph[now_idx][neighbor] < costs[neighbor]:
            heappush(heap, (now_cost+graph[now_idx][neighbor]+1, neighbor))

print(costs[-1])
```

호기롭게 틀려버렸다..

<img width="1140" alt="image" src="https://github.com/user-attachments/assets/51f1ddf0-41df-456e-afbd-eff38d05e3ef">

<br><br><br><br><br><br>

왜 틀렸을까 곰곰히 생각해보다가 아래와 같은 반례를 생각해냈다.

```
5 4
4 5
3 4
2 3
1 2
```

첫 번째 주기에서는 3 -> 4초에 2번노드로 이동되었고,

두 번째 주기에서 6 -> 7 초에 3번 노드로 이동되었고,

세 번째 주기에서는 9 -> 10 초에 4번 노드로 이동되었고,

네 번째 주기인 12 -> 13 초가 되어서야 5번 노드로 도착할 수 있다.

즉, 한 주기가 지날 때 마다 **총 주기인 M을 고려해서 시간을 더해줘야 된다는 것**.

<br><br><br><br><br><br>

또 여기서 한 번 더 생각해줘야 하는 것이,

아래와 같은 예시가 있을 때,

```
5 5
4 5
3 4
2 3
1 2       // 3-> 4초에 2번 노드로 이동
2 3      // 4 -> 5초에 바로 3번 노드로 이동 가능
```

`2 -> 3` 간선이 여러개 있을 때, 2번노드로 갈 수 있는 간선이 해당 간선 이전에 등장할 경우 똑같은 주기에서 이동할 수 있다는 것이다.

이 것만 처리해주면 야무지게 점수를 챙길 수 있다!

**이제 이 것을 어떻게 처리해줄 것이냐가 관건!!!**


## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

H0ngJu commented on tgyuuAn/다익스트라/횡단보도.py:
```diff
@@ -0,0 +1,40 @@
+        
+        for neighbor_idx in graph[now_idx][neighbor]:
+    
+            real_idx = now_cost % M
+            if real_idx < neighbor_idx:
+                heappush(heap, (neighbor_idx - real_idx + now_cost, neighbor))
+            else:
+                heappush(heap, (now_cost + M - real_idx + neighbor_idx, neighbor))
```
PR 마지막 케이스 경우,
2->3 으로 가는 간선이 2개니까 
간선 index(=real_idx)보다 이웃 간선의 index가 크냐 작냐를 비교해서 다르게 업뎃하는거 맞나예?!

tgyuuAn replied:
맞습니뎨!!!!!

H0ngJu reviewed (approved):
흐어 기본 다익스트라에서 변형할게 너무 많아서 까다로웠습니다ㅏ..

결국 다른 레퍼런스들 참고해ㅆ습니다요 ...ㅎ;ㅎ;
다익스트라 문제를 더 풀어봐야겠다는 마음을 먹고 갑니다 ,,,,,,,,

고생하셨슴다 ~~

```python
import sys
import heapq

def input(): return sys.stdin.readline().strip()

N, M = map(int, input().split())

INF = int(1e18)

graph = [[] for _ in range(N+1)]
cost = [INF] * (N+1)

for i in range(M):
    u, v = map(int, input().split())
    graph[u].append((v, i))
    graph[v].append((u, i))

def dijkstra():
    q = []
    heapq.heappush(q, (0, 1))
    cost[1] = 0
    
    while q:
        time, now = heapq.heappop(q)

        if cost[now] < time:
            continue
        
        for next, value in graph[now]:
            if time % M > value:
                    next_time = time + M - (time % M - value)
            else: 
                next_time = time + value - time % M

            if cost[next] > next_time+1:
                cost[next] = next_time+1
                heapq.heappush(q, (next_time+1, next))

    return cost[N]

print(dijkstra())
```

tgyuuAn commented:
> 흐어 기본 다익스트라에서 변형할게 너무 많아서 까다로웠습니다ㅏ..
> 
> 
> 
> 결국 다른 레퍼런스들 참고해ㅆ습니다요 ...ㅎ;ㅎ;
> 
> 다익스트라 문제를 더 풀어봐야겠다는 마음을 먹고 갑니다 ,,,,,,,,
> 
> 
> 
> 고생하셨슴다 ~~
> 
> 
> 
> ```python
> 
> import sys
> 
> import heapq
> 
> 
> 
> def input(): return sys.stdin.readline().strip()
> 
> 
> 
> N, M = map(int, input().split())
> 
> 
> 
> INF = int(1e18)
> 
> 
> 
> graph = [[] for _ in range(N+1)]
> 
> cost = [INF] * (N+1)
> 
> 
> 
> for i in range(M):
> 
>     u, v = map(int, input().split())
> 
>     graph[u].append((v, i))
> 
>     graph[v].append((u, i))
> 
> 
> 
> def dijkstra():
> 
>     q = []
> 
>     heapq.heappush(q, (0, 1))
> 
>     cost[1] = 0
> 
>     
> 
>     while q:
> 
>         time, now = heapq.heappop(q)
> 
> 
> 
>         if cost[now] < time:
> 
>             continue
> 
>         
> 
>         for next, value in graph[now]:
> 
>             if time % M > value:
> 
>                     next_time = time + M - (time % M - value)
> 
>             else: 
> 
>                 next_time = time + value - time % M
> 
> 
> 
>             if cost[next] > next_time+1:
> 
>                 cost[next] = next_time+1
> 
>                 heapq.heappush(q, (next_time+1, next))
> 
> 
> 
>     return cost[N]
> 
> 
> 
> print(dijkstra())
> 
> ```

와 이거 어려운뭄젠데 짜란다 짜라낟 짜란다~

alstjr7437 commented on tgyuuAn/다익스트라/횡단보도.py:
```diff
@@ -0,0 +1,40 @@
+    graph[start][destination].append(idx)
+    graph[destination][start].append(idx)
+    
+costs = [MAX for _ in range(N+1)]
```
이부분에서 sys.maxsize를 안하고 따로 선언한건 문제에 최대값까지 지정하기 위해서인가요??

tgyuuAn replied:
마아아아앚습니다! 사실 `int(1e9)` 해도 상관없어요

alstjr7437 commented:
문제를 풀 시간이 많지 않아서 코드는 작성을 못해보고 어떠한 로직인지만 살펴봤는데

다익스트라에서 간선의 인덱스로 현재 비용과 잘 섞어서 풀어야 겠네요..

따로 시간초과 부분을 처음에 생각하고 문제를 푸시는 것 같은데 알고리즘 문제 풀때
문제 확인 -> 시간 복잡도 계산 -> 알고리즘 결정 -> 데이터 타입? 결정 -> 구현 
이런식으로 푸시나요??

tgyuuAn commented:
> 문제를 풀 시간이 많지 않아서 코드는 작성을 못해보고 어떠한 로직인지만 살펴봤는데
> 
> 다익스트라에서 간선의 인덱스로 현재 비용과 잘 섞어서 풀어야 겠네요..
> 
> 따로 시간초과 부분을 처음에 생각하고 문제를 푸시는 것 같은데 알고리즘 문제 풀때 문제 확인 -> 시간 복잡도 계산 -> 알고리즘 결정 -> 데이터 타입? 결정 -> 구현 이런식으로 푸시나요??

어떤 문제든 제한 사항을 먼저확인하고,

제한이 빡빡한 것 같다 싶으면 바로 시간복잡도 계산을 먼저 해봅니다!!

근데 이제 문제에서 메모리를 지나치게 많이준다, 그러면 DP혹은 그래프 이런쪽으로도 생각해봅니다!
